crates/dataflow-jit/src/codegen/json/deserialize.rs

```diff
@@ -252,30 +252,130 @@ impl Codegen {
                         }
                     }
 
-                    ty @ (ColumnType::Date | ColumnType::Timestamp) => {
-                        let intrinsic = match ty {
-                            ColumnType::Date => "deserialize_json_date",
-                            ColumnType::Timestamp => "deserialize_json_timestamp",
-                            _ => unreachable!(),
+                    ColumnType::Date => {
+                        let spec = json_column
+                            .spec()
+                            .expect("dates require a json parsing spec");
+
+                        let value_is_null = match spec {
+                            JsonColumnParseSpec::DateTimeFromStr { format } => {
+                                let deserialize = ctx.imports.get(
+                                    "deserialize_json_date",
+                                    ctx.module,
+                                    builder.func,
+                                );
+
+                                let (format_ptr, format_len) =
+                                    ctx.import_string(format.to_string(), &mut builder);
+
+                                builder.call_fn(
+                                    deserialize,
+                                    &[
+                                        column_place,
+                                        json_pointer,
+                                        json_pointer_len,
+                                        format_ptr,
+                                        format_len,
+                                        json_map,
+                                    ],
+                                )
+                            }
+
+                            JsonColumnParseSpec::DateFromDays => {
+                                let deserialize = ctx.imports.get(
+                                    "deserialize_json_date_from_days",
+                                    ctx.module,
+                                    builder.func,
+                                );
+
+                                builder.call_fn(
+                                    deserialize,
+                                    &[column_place, json_pointer, json_pointer_len, json_map],
+                                )
+                            }
+
+                            JsonColumnParseSpec::TimeFromMicros
+                            | JsonColumnParseSpec::TimeFromMillis => unreachable!(),
```

Comment:
an error message would make it much easier to diagnose misconfigurations.
e.g., "configuration X not supported for type Y"